workflow/sync/multi_throttler.go

```diff
@@ -0,0 +1,243 @@
+func NewMultiThrottler(parallelism int, namespaceParallelism map[string]int, namespaceParallelismDefault int, queue QueueFunc) Throttler {
+	return &multiThrottler{
+		queue:                       queue,
+		namespaceParallelism:        namespaceParallelism,
+		namespaceParallelismDefault: namespaceParallelismDefault,
+		totalParallelism:            parallelism,
+		running:                     make(map[Key]bool),
+		pending:                     make(map[string]*priorityQueue),
+		lock:                        &sync.Mutex{},
+	}
+}
+
+type multiThrottler struct {
+	queue                       QueueFunc
+	namespaceParallelism        map[string]int
+	namespaceParallelismDefault int
+	totalParallelism            int
+	running                     map[Key]bool
+	pending                     map[string]*priorityQueue
+	lock                        *sync.Mutex
+}
+
+func (m *multiThrottler) Init(wfs []wfv1.Workflow) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	keys := []Key{}
+	for _, wf := range wfs {
+		if wf.Status.Phase != wfv1.WorkflowRunning {
+			continue
+		}
+		key, err := cache.MetaNamespaceKeyFunc(&wf)
+		if err != nil {
+			return err
+		}
+		keys = append(keys, key)
+	}
+
+	for _, key := range keys {
+		m.running[key] = true
+	}
+	return nil
+}
+
+func (m *multiThrottler) namespaceCount(namespace string) (int, int) {
+	setLimit, has := m.namespaceParallelism[namespace]
+	if !has {
+		m.namespaceParallelism[namespace] = m.namespaceParallelismDefault
+		setLimit = m.namespaceParallelismDefault
+	}
+	count := 0
+	for key := range m.running {
+		ns, _, _ := cache.SplitMetaNamespaceKey(key)
+		if ns == namespace {
+			count++
+		}
+	}
+	return count, setLimit
+}
+
+func (m *multiThrottler) namespaceAllows(namespace string) bool {
+	count, limit := m.namespaceCount(namespace)
+	return count < limit || limit == 0
+}
+
+func (m *multiThrottler) Add(key Key, priority int32, creationTime time.Time) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	namespace, _, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return
+	}
+	_, ok := m.pending[namespace]
+	if !ok {
+		m.pending[namespace] = &priorityQueue{itemByKey: make(map[string]*item)}
+	}
+
+	m.pending[namespace].add(key, priority, creationTime)
+	m.queueThrottled()
+}
+
+func (m *multiThrottler) Admit(key Key) bool {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, ok := m.running[key]
+	if ok {
+		return true
+	}
+	m.queueThrottled()
+	return false
+}
+
+func (m *multiThrottler) Remove(key Key) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	namespace, _, _ := cache.SplitMetaNamespaceKey(key)
+	delete(m.running, key)
+	m.pending[namespace].remove(key)
+	m.queueThrottled()
+}
+
+func (m *multiThrottler) queueThrottled() {
+	if m.totalParallelism != 0 && len(m.running) >= m.totalParallelism {
+		return
+	}
+
+	minPq := &priorityQueue{itemByKey: make(map[string]*item)}
+
+	cnts := make(map[string]int)
```

Comment:
This dubiously named variable appears to be write only, remove?

Reply:
hahahaha, yes I meant to remove this, thanks.